src/test/rgw/test_rgw_lc.cc

```diff
@@ -106,3 +106,238 @@ TEST(TestLCFilterInvalidAnd, XMLDoc3)
   /* check our flags */
   ASSERT_EQ(filter.get_flags(), uint32_t(LCFlagType::none));
 }
+
+struct LCWorkTimeTests : ::testing::Test
+{
+   CephContext* cct;
+   std::unique_ptr<RGWLC::LCWorker> worker;
+
+   // expects input in the form of "%m/%d/%y %H:%M:%S"; e.g., "01/15/23 23:59:01"
+   utime_t get_utime_by_date_time_string(const std::string& date_time_str)
+   {
+      struct tm tm{};
+      struct timespec ts = {0};
+
+      strptime(date_time_str.c_str(), "%m/%d/%y %H:%M:%S", &tm);
+      ts.tv_sec = mktime(&tm);
+
+      return utime_t(ts);
+   }
+
+   // expects a map from input value (date & time string) to expected result (boolean)
+   void run_should_work_test(const auto& test_values_to_expectations_map) {
+      for (const auto& [date_time_str, expected_value] : test_values_to_expectations_map) {
+         auto ut = get_utime_by_date_time_string(date_time_str);
+         auto should_work = worker->should_work(ut);
+
+         ASSERT_EQ(should_work, expected_value)
+            << "input time: " << ut
```

Comment:
I don't think I've seen this;  ASSERT* yield an outputstream?

Reply:
Ahh yes, this is a nice little feature gtest has: https://github.com/google/googletest/blob/main/docs/reference/assertions.md. Assertion macros support streaming custom messages into them. The nice thing is the custom message is printed only when assertion fails. For example, just changed a testcase to fail to see the custom output:

<pre>
[ RUN      ] LCWorkTimeTests.ShouldWorkCustomWorkTimeStartEndSameHour
[       OK ] LCWorkTimeTests.ShouldWorkCustomWorkTimeStartEndSameHour (3 ms)
[ RUN      ] LCWorkTimeTests.ScheduleNextStartTime
../src/test/rgw/test_rgw_lc.cc:148: Failure
Expected equality of these values:
  wait_secs_till_next_start
    Which is: 75713
  expected_value
    Which is: 75700
work_started_at: 2023-01-01T22:15:05.000000-0500 work_completed_at: 2023-01-01T01:13:07.000000-0500 expected: 75700 wait_secs_till_next_start: 75713 work-time-window: 22:15-03:33

[  FAILED  ] LCWorkTimeTests.ScheduleNextStartTime (3 ms)
[----------] 8 tests from LCWorkTimeTests (22 ms total)

[----------] Global test environment tear-down
[==========] 11 tests from 3 test suites ran. (23 ms total)
[  PASSED  ] 10 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] LCWorkTimeTests.ScheduleNextStartTime

 1 FAILED TEST
</pre>

Here, `work_started_at: 2023-01-01T22:15:05.000000-0500 work_completed_at: 2023-01-01T01:13:07.000000-0500 expected: 75700 wait_secs_till_next_start: 75713 work-time-window: 22:15-03:33` is the custom log line streamed into assertion. The cases that pass hide this extra data.